### DISCUSSION TITLE
Lighthouse: Remove unused JavaScript

### DISCUSSION THREAD

**Discussion by enricoschaaf** [vercel/next.js#13682] [Help] [answered]
I build this little polling site  https://poll-hp61gmonn.now.sh , but when I test the site with Lighthouse (https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fpoll-hp61gmonn.now.sh%2F) I get the message that I ship unused javascript. I analized by code using Madge but I havent found dead code this way. Could it be that one of the libaries I use is causing this?

My dependencies are:

    "@prisma/client": "^2.0.0-beta.7", //api-route
    "framer-motion": "^2.0.0-beta.72",
    "nanoid": "^3.1.9",
    "next": "^9.4.4",
    "preact": "^10.4.4",
    "preact-render-to-string": "^5.1.9",
    "react": "github:preact-compat/react#1.0.0",
    "react-dom": "github:preact-compat/react-dom#1.0.0",
    "react-hook-form": "^5.7.2",
    "swr": "^0.2.2"

The repo can be found here: https://github.com/enricoschaaf/poll

Thanks a lot!

Edit: I now created a fresh project with the vercel Next.js template and I get the same message. Is there currently a problem with bundling in Next.js?

Edit 2: Just reran lighthouse on the "Hello World" Next.js project. Now I dont get any message. Maybe there is a problem with lighthouse.

**Comment by Vadorequest** [comment]
I have a similar issue, explained there:
https://github.com/vercel/next.js/discussions/13646

**Reply by enricoschaaf** [reply]
I am basically getting remove unused JavaScript with a npx create-next-app, I think it's a problem with Lighthouse, if I run Lighthouse on an older version, I don't get this message.

**Comment by brianlovin** [comment]
I am seeing this on my end as well today - lighthouse scores for various pages on my website dropped from ~90 to ~75.

**Comment by Vadorequest** [comment]
I opened an issue on LightHouse.

https://github.com/GoogleChrome/lighthouse/issues/10904

**Comment by patrickhulce** [comment, marked as answer]
Hey everyone Lighthouse maintainer here 👋 

This is working as intended in Lighthouse. It is a new audit in 6.0 that highlights files that have a considerable amount of unused JavaScript. In that URL that's highlighted, about 70% of the main chunk is unused. You can explore each line in detail using the "Coverage" panel in DevTools.

![screenshot of the coverage panel highlighting unused portions of the main chunk related to animations](https://user-images.githubusercontent.com/2301202/83658498-e63ac280-a587-11ea-995c-16a04dca9608.png)

*screenshot of the coverage panel highlighting unused portions of the main chunk related to animations*

It is somewhat inevitable that some subset of JavaScript will be unused when using someone's library. They will have built features that you simply don't need. It's not important to obsess over 100% usage as that will be virtually impossible, but when it's a rather large amount of unused bytes, it's worth taking a look to see what you might be able to split out or if you really need the library.

Also worth noting that this new audit highlighting unused javascript in no way impacts the Lighthouse score. It is only suggesting one potential approach to help improve performance. Opportunities never impact the score and the score changes are a result of the new metrics that were added in 6.0, LCP and TBT.

**Reply by enricoschaaf** [reply]
Nice, thanks for the detailed answer!

**Reply by Vadorequest** [reply]
Indeed, thanks a lot!

> It is somewhat inevitable that some subset of JavaScript will be unused when using someone's library.

Agreed. I haven't given a try using a brand new Next.js app, but I've seen reports stating the results weren't good either.

@timneutkens Maybe is there room for improvement on Next.js so that a brand new app uses most of its code? I'm not familiar with that kind of optimisations. Splitting the app in more smaller bundles would increase network ops, so it's hard to tell beforehand if that would actually improve performances.

**Reply by timneutkens** [reply]
Next.js already optimized chunk splitting in 9.2 in collaboration with the Chrome team: https://nextjs.org/blog/next-9-2#improved-code-splitting-strategy. We're always optimizing more, however it's likely that in this case it's actually user code so can you provide a full reproduction?

**Reply by Vadorequest** [reply]
What would you need for reproduction? 

I've got a branch at https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-gcms-lcz-sty

LightHouse report at https://lighthouse-dot-webdotdevsite.appspot.com//lh/html?url=https%3A%2F%2Fnrn-v2-mst-aptd-gcms-lcz-sty-c1-5y65ioi14.now.sh%2Ffr

Based on https://nrn-v2-mst-aptd-gcms-lcz-sty-c1-5y65ioi14.now.sh/fr demo

Is that enough or do you need anything else?

**Reply by Ollynov** [reply]
@timneutkens Just confirming that even when running the official example blog from the Nextjs documentation, it still gives this warning. It is still a Performance score of 99, so still awesome, but just a heads up :) 

here is a screenshot in case it helps
![Screen Shot 2020-08-09 at 10 24 10 PM](https://user-images.githubusercontent.com/17999489/89753983-12deef00-da8f-11ea-8b4d-212672e83e6c.png)

**Reply by zoltanradics** [reply]
@Ollynov `chrome-extension://` line indicates that you are running Lighthouse in an environment where your Chrome extensions are loaded. Do the Lighthouse audit in Incognito mode and make sure you don't load any extension in Incognito mode.

**Reply by SalahAdDin** [reply]
@zoltanradics You made me see the light.

**Reply by arsnl** [reply]
@patrickhulce A question still. Why Lighthouse include the chrome-extension files into the analysis and not only the targeted Website files? chrome-extension files are not part of the Website but part of the user browser.

**Reply by patrickhulce** [reply]
@arsnl 

1. When a website is being analyzed with Chrome extensions that are injecting scripts, it is impossible to dissect what was the page and what was the Chrome extension (not all extension activity will have a clean stack trace). Sweeping Chrome extension activity under the rug would be disingenuous and make it even harder to tell why perf is so much worse than another environment without the extension. Lighthouse always reports what it observed for consistency, Chrome extensions, anti-virus, corp policy injections, whatever else the machine has going on, etc.
2. Many Chrome extension authors actually use this to measure their impact on pages.

**Reply by zoltanradics** [reply]
@arsnl Yes indeed, Lighthouse should exclude from the analysis the JS files loaded from `chrome-extension://` or at least give us possibilty to opt-in if we want to include them or not. But this isn't the case so until that just make sure you are running Lighthouse within a clean incognito mode or keep a clean Chromium installed on your computer.

**Reply by jserrao** [reply]
@patrickhulce - not feeling great that your service constantly hammers my Next sites into the 60s when they load extremely well on  every device I test on. You all really need to rethink what you're doing over there.

**Reply by ryanwalters** [reply]
Not sure if this question is more for @patrickhulce or @timneutkens, but how does `<Link>` prefetching get factored into the PageSpeed score?

For example, on one of our Next.js sites, when testing the home page in PageSpeed Insights console, one of the primary offenders is the JS chunk for a different page (e.g. _next/static/chunks/pages/%5BquestionSlug%5D-fa20d71c82a35ec047e6.js). I'm guessing this is due to `<Link>` prefetching that page, since we link to it from the home page.

I would have assumed that `<Link>` prefetching is deferred until after the rest of the page is loaded, or [not at all](https://web.dev/route-prefetching-in-nextjs/#how-automatic-prefetching-works) if prefetching is disabled when connection speeds are slow.

**Reply by JonCognioDigital** [reply]
Why is @jserrao getting downvoted when he makes a very valid point?

My site is built using NextJS, hosted on Vercel and the user experience is lightning fast (loads and is interactive in way under a second), so why is it OK that Lighthouse gives it a terrible score and presumably it gets demoted by Google in search results because of it's perceived slow performance even though this is not indicative of real life performance?

On another note, is there something that the Next team can do to improve the Lighthouse performance scores fo NExtJS sites out of the box? All I can see in lighthouse is lots of files ending in chunk.js which obviously we have no control over because the framework produced them and all of them are showing up as "unused javascript" (even though they're clearly not).

**Reply by realestninja** [reply]
2 years later and this issue is still present. I'm manually opting out of the nextjs `Link` pre-fetching for viewport components that cause the pre-fetch which lighthouse doesn't seem to like.

Does anyone have a better idea?

**Reply by SalahAdDin** [reply]
No any solution yet?

**Reply by yuvalkarmi** [reply]
This is absolutely bonkers. I spent a week optimizing my NextJS site only to discover that the primary culprit is the <Link> tag prefetching pages. That being said - this *does* affect the user experience - even an async chunk that's fetched gets parsed as soon as it's done downloading, and this *can* block the main thread while it's still rendering the page (low CPU, high enough network speed). There's not way to "defer" these chunks in next, but as an experiment, if you put all of those scripts in a setTimeout(..., 3000), then the page shows up INSTANTLY and the lighthouse score is suddenly lighting fast.

**Comment by titoabreurj** [comment]
Hi, 
I would like a Plugin that would actually do that on my website. 

Would anyone have a Plugin to indicate?

Need a plugin that really tires the css and th JS unused on the page. 
That I can fulfill LightHouse's requests. 
I'm using Wordpress, currently the Plugin wich is W3 Total Cache.

Thanks for all
Tito
